elk/evaluation/evaluate.py

```diff
@@ -1,121 +1,78 @@
+            `elk.training.preprocessing.normalize()` for details.
+        num_gpus: The number of GPUs to use. Defaults to -1, which means
+            "use all available GPUs".
+        debug: When in debug mode, a useful log file is saved to the memorably-named
+            output directory. Defaults to False.
+    """
+
+    data: Extract
     source: str = field(positional=True)
     normalization: Literal["legacy", "none", "elementwise", "meanonly"] = "meanonly"
-    num_gpus: int = -1
-
-
-def evaluate_reporter(
-    cfg: EvaluateConfig,
-    dataset: DatasetDict,
-    layer: int,
-    devices: list[str],
-    world_size: int = 1,
-):
-    """Evaluate a single reporter on a single layer."""
-    rank = os.getpid() % world_size
-    device = devices[rank]
-
-    # Note: currently we're just upcasting to float32 so we don't have to deal with
-    # grad scaling (which isn't supported for LBFGS), while the hidden states are
-    # saved in float16 to save disk space. In the future we could try to use mixed
-    # precision training in at least some cases.
-    with dataset.formatted_as("torch", device=device, dtype=torch.int16):
-        train_split, val_split = select_train_val_splits(dataset)
-        train, val = dataset[train_split], dataset[val_split]
-        test_labels = cast(Tensor, val["label"])
-
-        _, test_h = normalize(
-            int16_to_float32(assert_type(Tensor, train[f"hidden_{layer}"])),
-            int16_to_float32(assert_type(Tensor, val[f"hidden_{layer}"])),
-            method=cfg.normalization,
-        )
-
-    reporter_path = elk_reporter_dir() / cfg.source / "reporters" / f"layer_{layer}.pt"
-    reporter = torch.load(reporter_path, map_location=device)
-    reporter.eval()
 
-    test_x0, test_x1 = test_h.unbind(dim=-2)
+    debug: bool = False
+    out_dir: Optional[Path] = None
+    num_gpus: int = -1
 
-    test_result = reporter.score(test_labels, test_x0, test_x1)
+    def execute(self):
+        transfer_eval = elk_reporter_dir() / self.source / "transfer_eval"
+        cols = ["layer", "loss", "acc", "cal_acc", "auroc"]
 
-    stats = [layer, *test_result]
-    return stats
+        run = Evaluate(cfg=self, eval_headers=cols, out_dir=transfer_eval)
+        run.evaluate()
 
 
-def evaluate_reporters(cfg: EvaluateConfig, out_dir: Optional[Path] = None):
-    ds = extract(cfg.target, num_gpus=cfg.num_gpus)
+@dataclass
+class Evaluate(Run):
+    cfg: Eval
 
-    layers = [
-        int(feat[len("hidden_") :])
-        for feat in ds["train"].features
-        if feat.startswith("hidden_")
-    ]
+    def evaluate_reporter(self, layer: int, devices: list[str], world_size: int):
+        """Evaluate a single reporter on a single layer."""
+        device = self.get_device(devices, world_size)
 
-    devices = select_usable_devices(cfg.num_gpus)
-    num_devices = len(devices)
+        _, _, test_x0, test_x1, _, test_labels = self.prepare_data(
```

Comment:
I think we should split up the extraction of train and test data so we don't have to do this